# Homework 3

In [289]:
import pandas as pd
import numpy as np
import plotly.graph_objects as go
import plotly.express as px
from tqdm import tqdm 
import plotly.figure_factory as ff

##  Problem 1

In [3]:
fiscal_data=pd.read_csv('Sdf16_1a.txt',sep='\t',header=0,low_memory=False)
math_performance_data=pd.read_csv('math-achievement-lea-sy2015-16.csv',low_memory=False)
RLA_performance_data=pd.read_csv('rla-achievement-lea-sy2015-16.csv',low_memory=False)
fiscal_data.describe()

FIPST     YEAR         CCDNF       CENFILE            V33  \
count  18708.000000  18708.0  18708.000000  18708.000000   18708.000000   
mean      29.495029     16.0      0.998396      0.786402    2683.242196   
std       14.742041      0.0      0.040014      0.409857   12665.966250   
min        1.000000     16.0      0.000000      0.000000      -3.000000   
25%       18.000000     16.0      1.000000      1.000000     182.000000   
50%       30.000000     16.0      1.000000      1.000000     641.000000   
75%       40.000000     16.0      1.000000      1.000000    2003.000000   
max       56.000000     16.0      1.000000      1.000000  981667.000000   

           MEMBERSCH      TOTALREV       TFEDREV           C14           C15  \
count   18708.000000  1.870800e+04  1.870800e+04  1.870800e+04  1.870800e+04   
mean     2667.760049  3.731826e+07  2.972137e+06  7.488747e+05  5.972777e+05   
std     12639.367419  2.439561e+08  1.945279e+07  6.657532e+06  3.305317e+06   
min        -9.000000 -2.000000e+00 -2.000000e+00 -2.000000e+00 -2.000000e+00   
25%       181.000000  2.822000e+06  1.537500e+05  1.400000e+04  0.000000e+00   
50%       630.000000  9.177500e+06  5.760000e+05  1.260000e+05  5.800000e+04   
75%      1972.000000  2.885050e+07  1.815000e+06  4.160000e+05  3.730000e+05   
max    981290.000000  2.744836e+10  1.739101e+09  6.562260e+08  2.910520e+08   

       ...          _66V           W01           W31           W61  \
count  ...  1.870800e+04  1.870800e+04  1.870800e+04  1.870800e+04   
mean   ...  4.418570e+05  1.101326e+06  3.084204e+06  8.197737e+06   
std    ...  7.292256e+06  9.463948e+06  2.173486e+07  3.382881e+07   
min    ... -2.000000e+00 -2.000000e+00 -2.000000e+00 -2.000000e+00   
25%    ...  0.000000e+00  0.000000e+00  0.000000e+00  1.630000e+05   
50%    ...  0.000000e+00  0.000000e+00  0.000000e+00  1.854000e+06   
75%    ...  0.000000e+00  2.540000e+05  2.480000e+05  6.142500e+06   
max    ...  8.700000e+08  8.724840e+08  1.085345e+09  2.887956e+09   

                V95           V02           K14           CE1           CE2  \
count  1.870800e+04  1.870800e+04  1.870800e+04  1.870800e+04  1.870800e+04   
mean   3.298232e+05  1.770253e+05  4.803566e+04  1.258707e+07  1.199914e+06   
std    1.498774e+06  1.595693e+06  4.037505e+05  6.769750e+07  7.943880e+06   
min   -2.000000e+00 -2.000000e+00 -2.000000e+00 -2.000000e+00 -2.000000e+00   
25%    0.000000e+00  0.000000e+00  0.000000e+00 -1.000000e+00 -1.000000e+00   
50%    3.000000e+03  0.000000e+00  0.000000e+00 -1.000000e+00 -1.000000e+00   
75%    2.100000e+05  2.400000e+04  0.000000e+00  6.289000e+06  4.850000e+05   
max    7.966600e+07  1.556730e+08  1.577900e+07  2.768580e+09  4.273440e+08   

        WEIGHT  
count  18708.0  
mean       1.0  
std        0.0  
min        1.0  
25%        1.0  
50%        1.0  
75%        1.0  
max        1.0  

[8 rows x 132 columns]

In [336]:
state_fiscal_data=fiscal_data.groupby(['STABBR']).sum()
state_fedrevenue_data_ranked=state_fiscal_data[['TFEDREV','V33']].sort_values('TFEDREV',ascending=False).reset_index()
state_fedrevenue_data_ranked['fdfperstudent']=state_fedrevenue_data_ranked['TFEDREV']/state_fedrevenue_data_ranked['V33']
state_fedrevenue__by_student=state_fedrevenue_data_ranked[['STABBR','fdfperstudent']].sort_values('fdfperstudent',ascending=False).reset_index()
fig = go.Figure(data=go.Choropleth(
    locations=state_fedrevenue_data_ranked['STABBR'], # Spatial coordinates
    z = state_fedrevenue_data_ranked['TFEDREV'].astype(float), # Data to be color-coded
    locationmode = 'USA-states', # set of locations match entries in `locations`
    colorscale = 'Reds',
    colorbar_title = "Millions USD",
))

fig.update_layout(
    title_text = '2015-2016 Total Federal Revenue by State',
    geo_scope='usa', # limite map scope to USA
)

fig.show()

In [337]:
state_fedrevenue_data_ranked.iloc[0:10]

STABBR     TFEDREV      V33  fdfperstudent
0     CA  7709274795  6211175    1241.194266
1     TX  6194316950  5301386    1168.433491
2     NY  3374793578  2711496    1244.624214
3     FL  3147328988  2792233    1127.172764
4     IL  2334944853  2034632    1147.600575
5     PA  2037314939  1703465    1195.982858
6     OH  1837962930  1716407    1070.819992
7     GA  1815241986  1757035    1033.127960
8     MI  1731033937  1487192    1163.961302
9     NC  1587975932  1544858    1027.910612

In [338]:
state_fedrevenue__by_student.iloc[0:10]

index STABBR  fdfperstudent
0     41     DC    2713.262544
1     38     AK    2319.798908
2     12     LA    1552.709811
3     30     NM    1537.993342
4     42     MT    1518.538036
5     43     SD    1467.240384
6     39     HI    1434.825133
7     47     ND    1432.350143
8     26     MS    1417.876006
9     44     RI    1327.287060

In [339]:
fig = px.bar(state_fedrevenue_data_ranked, x='STABBR', y='TFEDREV')
fig.show()

## Problem 2

In [30]:
district_data=fiscal_data[['STABBR','LEAID','TOTALEXP','TOTALREV']].reset_index()
district_data['diff']=district_data['TOTALREV']-district_data['TOTALEXP']

In [31]:
labels=[]
for i in district_data['diff']:
    if i<0:
        labels.append(str('NEGATIve'))
    if i == 0:
        labels.append(str('EVEN'))
    if i > 0:
        labels.append(str("POSTIVE"))
        
district_data['labels']=labels


In [33]:
fig = px.scatter(district_data, x="TOTALREV", y="TOTALEXP",color='labels')
fig.show()

In [34]:
district_data.head()


index STABBR    LEAID   TOTALEXP   TOTALREV     diff    labels
0      0     AL  0100002         -1         -1        0      EVEN
1      1     AL  0100005   46778000   49795000  3017000   POSTIVE
2      2     AL  0100006   64739000   56422000 -8317000  NEGATIve
3      3     AL  0100007  165436000  168040000  2604000   POSTIVE
4      4     AL  0100008  101154000  105927000  4773000   POSTIVE

### STATE DEBT

In [40]:
state_data=district_data.groupby(['STABBR']).sum()
state_data[['diff']].sort_values('diff',ascending=True).reset_index()
state_data.iloc[0:8]

index     TOTALEXP     TOTALREV        diff
STABBR                                               
AK        11151   2623014000   2494691000  -128323000
AL        16110   7864635916   7607097916  -257538000
AR       316899   5544519973   5517203973   -27316000
AZ       417835   9482834922   9980176922   497342000
CA      2131220  85327524795  89224003795  3896479000
CO       672570  10020496867  10260557867   240061000
CT       579975  10985761984  11552644984   566883000
DC       219743   2277770985   2170631985  -107139000

## Problem 3

In [251]:
def unblur(column, delim='-'):
    replacement_indexs=[]
    for i in range(len(column)):
        if type(column[i]) == float :
            x=np.random.random_sample()*100
            replacement_indexs.append(x)
        else:
            stringvalues=list(column[i])
            if len(stringvalues)==1:
                if stringvalues[0]=='.':
                    x=np.random.random_sample()*100
                    replacement_indexs.append(x)
                else:
                    replacement_indexs.append(int(stringvalues[0]))
            elif len(stringvalues)==2:
                x=np.random.normal(50, 10, 1)[0]
                replacement_indexs.append(x)
            elif len(stringvalues)==3:
                if stringvalues[0]=='G' and stringvalues[1]=='E':
                    value=int(stringvalues[2])
                    x=np.random.randint(value,100, size=1)[0]
                    replacement_indexs.append(x)
                if stringvalues[0]=='G' and stringvalues[1]=='T':
                    value=int(stringvalues[2])
                    x=np.random.randint(value+1,100, size=1)[0]
                    replacement_indexs.append(x)
                elif stringvalues[0]=='L' and stringvalues[1]=='E':
                    value=int(stringvalues[2])
                    x=np.random.randint(value, size=1)[0]
                    replacement_indexs.append(x)
                elif stringvalues[0]=='L' and stringvalues[1]=='T':
                    value=int(stringvalues[2])
                    x=np.random.randint(1,value-1, size=1)[0]
                    replacement_indexs.append(x)
                else:
                    x=np.random.randint(int(stringvalues[0]),int(stringvalues[2]), size=1)[0]
                    replacement_indexs.append(x)
            elif len(stringvalues)==4:
                if stringvalues[0]=='G' and stringvalues[1]=='E':
                    value=int(stringvalues[2]+stringvalues[3])
                    x=np.random.randint(value,100, size=1)[0]
                    replacement_indexs.append(x)
                if stringvalues[0]=='L' and stringvalues[1]=='E':
                    value=int(stringvalues[2]+stringvalues[3])
                    x=np.random.randint(value, size=1)[0]
                    replacement_indexs.append(x)
                if stringvalues[0]=='G' and stringvalues[1]=='T':
                    value=int(stringvalues[2]+stringvalues[3])
                    x=np.random.randint(value+1,100, size=1)[0]
                    replacement_indexs.append(x)
                if stringvalues[0]=='L' and stringvalues[1]=='T':
                    value=int(stringvalues[2]+stringvalues[3])
                    x=np.random.randint(1,value-1, size=1)[0]
                    replacement_indexs.append(x)
            elif len(stringvalues)==5:
                value1=int(stringvalues[0]+stringvalues[1])
                value2=int(stringvalues[3]+stringvalues[4])
                x=np.random.randint(value1,value2, size=1)[0]
                replacement_indexs.append(x)
            else:
                print(stringvalues)
  
    return(replacement_indexs)
            
            

In [253]:
math_performance_data['good_Data']=unblur(math_performance_data['MAM_MTH00PCTPROF_1516'])

In [262]:
fig = px.histogram(math_performance_data, x="good_Data")
fig.update_layout(
    title="Percentage Math Proficent Histogram",
    xaxis_title="Percent of district that is Math Proficient",
    yaxis_title="Number of Districts",
    font=dict(
        family="Courier New, monospace",
        size=18,
        color="#7f7f7f"
    )
)
fig.show()

## Problem 4

In [265]:
fed_budget=fiscal_data['TFEDREV'].sum()

In [268]:
print(fed_budget)

55602739138


In [294]:
new_budget=.15*fed_budget

In [296]:
19121193000-new_budget

10780782129.3

In [295]:
print(new_budget)

8340410870.7


    The Total federal budget is 55602739138. To reduce by 15%, the new fed budget must be equal to: 47262328267.

In [324]:
district_data=fiscal_data[['STABBR','LEAID','TOTALEXP','TOTALREV','TFEDREV']].reset_index()
district_data['diff']=district_data['TOTALREV']-district_data['TOTALEXP']


In [320]:
def budget_reducer(data):
    school_ids=[]
    amount_reduced=[]
    for i in range(data.shape[0]):
        amount=data.iloc[i]['diff']
        if amount>0:
            school_ids.append(data.iloc[i]['LEAID'])
            if data.iloc[i]['diff']<=data.iloc[i]['TFEDREV']:
                amount_reduced.append(amount)
            else:
                amount_reduced.append(data.iloc[i]['TFEDREV'])
    fed_budget=data['TFEDREV'].sum()
    reduced_amount=.15*fed_budget
    current_reduced=sum(amount_reduced)
    while reduced_amount < current_reduced:
        current_reduced=sum(amount_reduced)
        if current_reduced>reduced_amount:
            reducedeach=(current_reduced-reduced_amount)/len(amount_reduced)
        for x in range(len(amount_reduced)):
            if amount_reduced[x] >= reducedeach:
                amount_reduced[x]=amount_reduced[x]-reducedeach
            else:
                amount_reduced[x]=amount_reduced[x]
        current_reduced=sum(amount_reduced)
    print(sum(amount_reduced))
    school_red=pd.DataFrame([school_ids,amount_reduced]).transpose()
    school_red.columns=['LEAID',"AMOUNT reduced"]
    return(school_red)

In [322]:
budget_reduction_table=budget_reducer(district_data)

8340410870.699973


## Problem 5

In order to reduce the Federal Budget by 15% we must make cuts to Federally funded programs. The following criteria are the rules by which funding was cut. 
        1.) Determine all districts that had a positive difference between their funding and their expenditure. 
        2.) Reduce district federal funding by the difference or by the federal funding whichever was smaller.
        3.) Determine whether the reduction was greater than needed. If so, iteratively reduce the reduction among all schools that had funding taken until the reduction equalled 15% of last year's budget.
        
This process was chose because it is clear, generally objective, and does not impact the learning of the schools that the money
is being taken from.
        